metrics/threshold_monitor.go

```diff
@@ -0,0 +1,105 @@
+package metrics
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/drand/drand/log"
+)
+
+type ThresholdMonitor struct {
+	lock              sync.RWMutex
+	log               log.Logger
+	beaconID          string
+	threshold         int
+	failedConnections map[string]bool
+	ctx               context.Context
+	cancel            func()
+	period            time.Duration
+}
+
+func NewThresholdMonitor(beaconID string, l log.Logger, threshold int) *ThresholdMonitor {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &ThresholdMonitor{
+		lock:              sync.RWMutex{},
+		log:               l,
+		beaconID:          beaconID,
+		threshold:         threshold,
+		failedConnections: make(map[string]bool),
+		ctx:               ctx,
+		cancel:            cancel,
+		period:            1 * time.Minute,
+	}
+}
+
+func (t *ThresholdMonitor) Start() {
+	t.log.Infow("starting threshold monitor", "beaconID", t.beaconID)
+
+	go func() {
+	FOR:
+		for {
+			select {
+			case <-t.ctx.Done():
+				t.log.Infow("ending threshold monitor", "beaconID", t.beaconID)
+				break FOR
+			default:
+				t.lock.RLock()
+				var failingNodes []string
+				for address := range t.failedConnections {
+					failingNodes = append(failingNodes, address)
+				}
+
+				if len(failingNodes) >= t.threshold {
+					t.log.Errorw(
+						"failed connections crossed threshold in the last minute",
+						"beaconID", t.beaconID,
+						"threshold", t.threshold,
+						"failures", len(failingNodes),
+						"nodes", strings.Join(failingNodes, ","),
+					)
+				} else if len(failingNodes) >= t.threshold/2 {
+					t.log.Warnw(
+						"failed connections crossed half threshold in the last minute",
```

Comment:
Same. Looks more like period given the sleep below to me

Reply:
the period here is different to the network period, by default it's 1min